keps/sig-node/3675-resource-plugin-manager/README.md

```diff
@@ -0,0 +1,1261 @@
+  and overriding current Kubelet allocation. We should provide a path otherwise. Many of these solutions are custom to particular companies. Many are not open source. The community should give a process to allow this functionality in core kubelet at a granular level, same as we have in many systems, such as HPC or telco.
+* Leverage additional parts from the pod spec information which could help make allocation decisions
+without the need to query the k8s controlplane.
+* Be able to do research, with minimum toil, on new policies and resource management strategies
+
+This design will also use gRPC, which is used for many other pluggable components within Kubernetes.  
+The hope is that it will also, as a side effect, allow a path to simplify the Kubelet as it is 
+today and move the complexity into external plugins.
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+#### Alpha Goals
+
+1.	Provide Initial infrastructure to support CCI Driver Plugins and k8s Deployments requiring CCI Drivers after scheduling
+2.	Support seamless k8s start ( no plugins required) through the intermediate CCI CPU management policy (functionality will be moved in beta to None and Static policies)
+3.	Support standard pod deployments not requiring CCI Drivers through CPU Manager: best-effort QoS(stretch goal will include static handling of burstable and guaranteed QoS)
+4.	CCI Driver results passed to CPU Manager (i.e Resource sets) have to support cpusets and memory affinity, .. 
+5.	Consistent on-none state handling between CCI Driver and CPU Manager (race-free)
+6.	Support proper fail mechanisms of Pods if CCI driver is not available - Error message + retry 
+7.	Continue cluster operation if CCI driver is not available
+8.  Podspecs:  Will be able to support current pod specs.  While there may be additional 
+extensibility in the future, the current pod specs will still work.
+
+
+#### Beta & Post-Beta Goals
+
+1.	Support standard pod deployments not requiring CCI Drivers through CPU & Memory Manager: all QoS types
+2.  Support hint providers for topology manager
+3.  Interoperability with Device Plugins, DRA, et cetera. 
+4.	Identify minimal in-cluster operational core (from existing CPU Manager, Memory Manager, Topology Manager)
+5.  Replace CCI policy with a component integrated in None and Static CPU Manager policies
+6.	E2E testing including other components
+
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+* Break any existing use cases for topology manager, memory manager, cpu manager, or 
+device manager.
+* Creating any more latency than there is today for scheduling:  We should be 
+careful not to add any latency into scheduling over what exists today for default behavior.
+* CPU & Memory Management on scheduler side.
```

Comment:
still not covered by CCI Drivers. But this is where we want to be :) . Most probably more work required

Reply:
Let's revisit this once comments from Kevin, which will require pivoting approach, are addressed. I'm still unconvinced yet we can implement an extensible approach without any change to resource representation, thus without any scheduler impact, and I'd love to zoom in here to understand how it could work.

Reply:
> Let's revisit this once comments from Kevin, which will require pivoting approach, are addressed. I'm still unconvinced yet we can implement an extensible approach without any change to resource representation, thus without any scheduler impact, and I'd love to zoom in here to understand how it could work.

I _strongly_ agree with this. Resource representation and scheduler impact is important and should be accessed.

Reply:
The way I see this working in the long-run is by reusing the `ResourceClass` and `ResourceClaim` mechanism added as part of the DRA work. Different ResourceClasses could be created to group CPUs and memory based on certain characteristics, and then users could request access to those via ResourceClaims.

It would require some small changes to the existing DRA API (e.g. to make sure we call out to CCI plugin and not a "normal" DRA plugin once we land on a node), but then we could basically reuse the same mechanisms we have today for coordinating allocations with the scheduler via a DRA controller, without needing to reinvent the wheel.

Reply:
thanks Kevin. This explanation is sufficient to answer my concerns. I'll need to review how DRA drives the scheduling, but that's totally on me (and very fair). I believe the KEP text as-is now emphasizes too strongly zero changes to the scheduling side and leave this implicit. I'd rather mention explicitely in the text, pretty much the way you described.

Reply:
Sure, `ResourceClass` and `ResourceClaim` mechanism makes sense when it comes to resource provisioning and consumption respectively. However, resource tracking and how scheduler becomes of aware of resources is not obvious to me. I understand that in DRA, we have a control plane driver component that takes responsibility of resource tracking and coordinating with the scheduler. Is the hardware vendor supposed to provide a DRA controller that works at a control plane level in addition to the CCIDriver which runs at a node level?  I haven't seen that mentioned in this KEP at all. It looks like we have some implicit assumptions and they probably need to be made explicite.

Reply:
As I said in the previous PR, it is not clear to me how a DRA driver could be written which reliably handles CPU allocation while still allowing kubelet+kube-scheduler to assign the same CPU resources to pods in the traditional way - it has to be either/or, not both.

So if such a solution is ultimately what this KEP is supposed to enable, then it should be in this KEP.